### ISSUE TITLE
Specify keep-alive and idle timeouts

### ISSUE THREAD

**Issue by MOZGIII** [w3c/webtransport#614]
The spec does not mention anything abut the keep-alive and idle timeouts.

The protocol itself permits setting protocol-level keep-alive, as well as the idle timeout. The application code should be able to, at the very least, request those parameters from the browser to learn what to expect, and ideally to request the parameters it would like during the transport setup.

I suggest allowing setting the optional keep-alive and idle timeouts, and throwing TypeError if they are not valid/acceptable. A reasonable fallback for the apps them would be to try negotiating the parameters by instantiating the WebTransport in a sequence, of attempts and catching the errors; with the ultimate fallback of not passing any keep-alive and idle timeout options.

---

Additions to WebTransport options:

- `keepaliveIntervalSeconds` (optional) - an duration of time, specified in seconds; if set, the browser will send a keel-alive packets to the server with this intervals. Can be used to keep idle but otherwise healthy connection from expiring.
- `idleTimeoutSeconds` (optional) - a duration of time, specified in seconds; specifies the browser side of idle timeout negotiation, the effective idle timeout if the minimum of durations suggested by the browser and the server; idle timeout specifies how long the WebTransport session will remain connected after the last activity it had; once the session becomes inactive, timeout errors will be emitted for all currently waiting operations.

Additions to the WebTransport API:

- `keepaliveIntervalSeconds` (property, `undefined | number`, readonly) - the effective keep-alive interval for this session.
- `idleTimeoutSeconds` (property, `number`, readonly) - the effective idle timeout for this session; 0 means no timeout.

**Comment by MOZGIII** [comment]
`fetch` has different API, and it is intended for request/response operations that are much shorter lived. Note that it could also benefit from the ability to tweak keepalive / idle parameters - although it is not so much of an issue there.

WebTransport could be used for a mobile game for instance, where the application would want to know and/or tweak the timings of the keepalive to allow the session to keep idle connections longer as users can move around.

Another example is an ngrok-like tunneling app, where the users and service providers might want to agree to let the session live for hours when idle, as this can be important for when the user moves locations (i.e. you go from the office to home while maintaining the same session).

**Comment by LPardue** [comment]
QUIC supports deferring idle timeout; see https://datatracker.ietf.org/doc/html/rfc9000#section-10.1.2. This is something applications need to drive themselves rather than the protocol itself running keepalive timers and proves as standard.

If I've learned any about timeouts in my space, it's that mixing Layer 4 and Layer 7 timeout concepts can cause confusion and operational issues. WebTransport is designed to work with intermediaries. I see it playing out a bit like websockets. Focusing on keeping a single hop QUIC or TCP connections alive between browser and frontend server might risk not addressing the needs of the application e.g the backend connection gets closed because there's no traffic travelling across. I've seen this first hand with gRPC too, where the (incorrect) expectation for H2 PING frames to travel end to end over several proxies doesn't happen.

For the OP use case, i think it _would_ be useful to allow setting a client idle timeout and for it to read the negotiated value that applies to the connection. However, I think then we need to place any timeout deferral into the responsibility of the application, by sending application data itself e.g a bidi stream that sends PING/PONG messages

**Comment by LPardue** [comment]
Right. But what I'm saying is,, I don't think the WebTransport API or it's implementations should spend effort trying to implement programmable hop-by-hop timeout avoidance. If they do, in my experience, app developers will still have to implement a real solution at their layer anyway.

I think we are agreed that setting/getting of the timeout values is useful.

**Comment by MOZGIII** [comment]
> I think we are agreed that setting/getting of the timeout values is useful.

👍 

Just to share some real-world application for context: I have a use-case that requires exactly hop-by-hop timeout avoidance - that is, allowing the *server* to go down and then up in couple hours while maintaining the same connection. The state of the server is preserved, only the link is cut - so what I, in fact, would like is to have a large enough idle timeout. It might not be what you mean by timeout avoidance (as in keep-alive) - but it is, in a way, timeout avoidance. So, just sharing my practical use case - where what I'd need is to have the idle timeout large enough with *no* keep-alive accompanying it. In fact, indeed, I *do* handle keep-alive at the application level - sort of, by avoiding sending anything that could trigger *a detection tat the link is lost*. So, it is odd, but very much intentional. My use case has to do with keeping the same connection for the cryptographic material and attesting the connection and remote system we are connected to.

**Comment by martinthomson** [comment]
Our implementation engages timeout avoidance when there is an open request.  WebTransport sessions count as open requests.

I'm not sure that there is much more to do than that.  Given the above, why would we need to give applications information about the idle timeout (which is what we use to determine the interval we use for sending keep-alive PING frames).

**Comment by LPardue** [comment]
In the discussed use case, the requirements are

1. Long idle timeout
2. No idle timeout avoidance via PINGs

That's because if the peer becomes uncontainable for a period, your PINGs will go unanswered and could end in the client giving up on the connection. Meanwhile the server springs back to life an holds state around for an idle client that is never going to come back.

But to answer the question, if we were to allow  the WebTransport API to set an client idle timeout value, the negotiated value might be smaller due to a server config. And if you build an application on top of WebTransport that is relying on certain idle timeout behaviour, ai suspect you'd want to know what the connection is actually using.

**Comment by martinthomson** [comment]
I'm not following that answer at all @LPardue.  If a packet isn't acknowledged, the client will very much treat that as an error and abort.  You can't assume that the browser won't be doing anything with a connection, such that a server can take a nap.

Recall that the server sets a value for idle timeout in transport parameters.  So we already have an API to set a value, it's just not JS (and it might not be end-to-end if you care about that; PING is not end-to-end either).  Presumably you will be able to find that value out too.

**Comment by LPardue** [comment]
A QUIC connection idle timeout is the min(client value, server value), so configuring your server to pick a large value doesn't mean much if a browser is picking something lower than you intend or desire.

Now, that aside, I completely concede the point that application code can't control everything a  QUIC connection might want or need to send. If its part of the "contract" that things can be sent underneath you even when you don't expect or want it to, that fact alone might make @MOZGIII 's use case impractical to implement. Examples include a server sending something ACK eliciting just before going to sleep, sending a STOP_SENDING that elicits a reset stream etc.

It might be that a server that goes to sleep is instructed on wakeup to immediately send an ACK-eliciting packet to quickly determine if the peer is still there or not, independent of any configured idle timeout.

In summary, while I do think the use case is valid and something that could work. It seems to be stretching the limits of what QUIC and WebTransport are really intended to support at present.

**Comment by MOZGIII** [comment]
> I completely concede the point that application code can't control everything a QUIC connection might want or need to send.

I do not disagree with this. At the same time, I would consider certain assumptions reasonable, under certain conditions.

This seems like you would be reasonable to assume certain emergent behavior of the transport when you request the browser to avoid transport connection reuse (or multiplexing multiple WebTransport requests over a single transport - i.e. QUIC - connection). Of course, unless, in practice, the browser is allowed to do keep-alive pings.

The capability to have a transport connection exclusive for your needs seems to actually be an exact fit for this - as in the ability to apply certain tweaks to the transport and assume additional constraints over how it would behave. Inability to control the transport-level pings kind of defeats the purpose - unless there is a different purpose for it. I can see, for instance, this as being a compatibility thing - not all servers support multiplexing of WebTransport sessions over a single transport connection.

> It might be that a server that goes to sleep is instructed on wakeup to immediately send an ACK-eliciting packet to quickly determine if the peer is still there or not, independent of any configured idle timeout.

That's actually what I'd do, yes. Also, in our case, the server wouldn't really go to sleep per se - it just disconnects from the network, for instance, changes the BGP routes to itself to blackhole. This would work alright with QUIC/UDP. Also, in our case, there is actual exact end-to-end connection, with respect to the server side, as this is always a concrete physical machine with a well-known software we're connecting to (in my case), not a cloud VM or something.



> In summary, while I do think the use case is valid and something that could work. It seems to be stretching the limits of what QUIC and WebTransport are really intended to support at present.

Exactly! I don't have high hopes for WebTransport to actually do something to support my use case. All I want is some clarity for myself and browser / server developers in the spec on whether it should be supported - or explicitly should not be.

In other words - currently we can inspect the browser implementations and make some decisions on our architecture based on how the implementations in the wild actually work. However, there is no guarantee they won't change their behavior, since it simply not specified. It would be nice if the spec actually clarified something. Like "assume browsers will cap idle timeouts" - that sort of clarification.

**Comment by LPardue** [comment]
To expand on my previous point, i think this is a QUIC level caveat not specific to WebTransport.

This is not just about connection pooling. The QUIC protocol requires that certain actions elicit a peer to respond. If you stop listening to the peer and don't acknowledge packets it sends, you are supposed to deal with that as a connection going dead. 

For example, the client might want to give more flow control to your server just as it becomes inaccessible to the network. Or it may have to respond to a server's STOP_SENDING with its own RESET_STREAM, just as it becomes inaccessible.

The QUIC protocol doesn't provide explicit support for what is being requested here. While it might work if everything lines up nicely, there's no requirement on implements of QUIC to do that.

**Comment by martinthomson** [comment]
I would expose this as `idleTimeout`.  The value is the minimum of the transport parameters set by client and server.  However, there is a problem in that this is only going to be available in HTTP/3, not HTTP/2.

**Comment by martinthomson** [comment]
Oh, it's worth observing that browsers do have idle timeouts for HTTP/2 as well.  They just can't know what the server value is.  So we could ask browsers to put an implementation-defined value in when using HTTP/2.

Another thing to note: there is also the risk that the server abandons the connection early, in addition to the network doing that, as was already mentioned.

**Comment by MOZGIII** [comment]
> Another thing to note: there is also the risk that the server abandons the connection early, in addition to the network doing that, as was already mentioned.

Yes, idle timeout is a very specific thing. Its value doesn't provide strong guarantees that the connection will survive while idle for the said duration of time - only that it will be for sure timed out after that duration...

---

...but at the same time it is a part of one of the most tricky and cumulatively complex parts of the networking. In my view going lower-lever and avoiding the over-abstraction here is the way to go. Maybe we can even provide a much more advanced API for querying transport-level properties - to adequately represent parts that are different for HTTP/2 and HTTP/3?

We can think about it like this: for some (maybe most) apps this APIs won't be useful; however, for other apps those API would be a necessity - but they would only reliably work if the data presented is unambiguous and accurate. Otherwise it risks to be one of those APIs that exists but no one can use due to nuances and lack of clarity.

**Comment by jan-ivar** [comment]
So sounds like there's some discussion needed still to determine whether exposing anything here is needed. As for a PR, I'm hearing

```js
const idleTimeout = wt.maxIdleTimeout; // the unambiguous QUIC value or impl-defined HTTP/2 value
```

**Comment by MOZGIII** [comment]
Counter proposal, more explicit APIs (not great, but just to make a point - as looks like my prev. message was not accounted for):
```
const idleTimeout = wt.maxQuicIdleTimeout; // the unambiguous QUIC value (or undefined)
// and
const idleTimeout = wt.maxHttp2IdleTimeout; // impl-defined HTTP/2 value (or undefined)
```
(or something better)

**Comment by MOZGIII** [comment]
Ok, what if:

```js
const idleTimeout = wt.getIdleTimeout(/* semantics */ "bestEffort");
```

(forgive me for using TypeScript, probably not the way used here to communicate things, but I'm familiar with it and now so much with WebIDL and it's still pretty intuitive)
```ts
type getIdleTimeout = (semantics?: Semantics = "bestEffort" ) => number;
type Semantics = "bestEffort" | "quicTransportMaxIdleTimeout" | "http2TransportImplementationDependent";
```

**Comment by MOZGIII** [comment]
> no max, we have enough of those already

Well, yes, max even doesn't make sense for *getting* the idle timeout - it makes sense only to *set* max idle timeout when creating the WebTransport; for the API that would give you the *effective* timeout - max would be misleading.

For the WebTransport construction API - the max is what quic would use, but not what HTTP2 would; so, maybe a `preferredIdleTimeout` would better capture the actual underlying semantics for the field? And then, if you'd check the `wt.getIdleTimeout` later (btw, should it be `Promise<number>` maybe? note for self to think about it later...) - you'd be able to compare the values and act (maybe give up on doing any further work and show the error to the user that this client was unable to negotiate the required idle timeout, etc - for some use cases) accordingly.

**Comment by martinthomson** [comment]
> `wt.getIdleTimeout(/* semantics */ "bestEffort");`

You can already see if the connection is using QUIC or HTTP/2, so you can infer which of these applies.  I don't see a lot of point to breaking this down.

**Comment by jan-ivar** [comment]
Meeting:
- No strong signal of demand from the group. If someone has a good use case for this feature we would like to hear from them. Otherwise we would likely not add this at this time.

**Comment by MOZGIII** [comment]
> someone
me..?

**Comment by jan-ivar** [comment]
Do you have a use case that is satisfied by https://github.com/w3c/webtransport/pull/685 better than trial-by-error (which would account for all hops not just the first)? Note that [max_idle_timeout](https://www.rfc-editor.org/rfc/rfc9000#section-10.1) seems to guarantee timeout after a certain time, not guarantee no timeout happens ahead of this time. Proposed use cases so far all seem to desire the latter guarantee.

**Comment by jan-ivar** [comment]
Meeting:
- The PR linked #685 does not solve the OP use case of allowing a server to nap. We don't know of a satisfactory solution to the OP problem.
- Closing based on previous comment. If a compelling use case is presented we can revisit this decision.
